---
title: "LR2TF-Usage"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{LR2TF-Usage}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

This is a guide for the usage of the LR2TF package. It is used to analyze scRNA-seq data for transcription factor activities. The transcription factor activities are estimated using the decoupleR tool. Using the DoRothEA regulon version from decoupleR and post-translational interactions from the Omnipath database[2,3,4], connections are made between transcription factors and ligands and receptors.

The results can be combined with ligand-receptor interactions and then analyzed using CrossTalkeR (https://github.com/CostaLab/CrossTalkeR/) [5].

## Installation
The package can be installed directly from github:
&nbsp;

```{r, include = TRUE, eval = FALSE}
library(devtools)
install_github("CostaLab/LR2TF")

library(remotes)
install_github("CostaLab/LR2TF", build_vignettes = TRUE)
```
&nbsp;

## Transcription factor activity predictions using decoupleR in python

For the transcription factor activity predictions we recommend using the decoupleR package in its python version. This provides a better performance, especially for larger data sets. To perform the predictions with decoupleR, we need an AnnData (.h5ad) file. If the scRNA-seq data is saved as a Seurat object (.Rds file), it can be converted to AnnData format using the sceasy package and the function we provide:

```{r, include = TRUE, eval = FALSE}
library(LR2TF)
library(Seurat)
seurat_object <- readRDS("/path/to/seurat_object.Rds")
LR2TF::convert_seurat_to_anndata(seurat_object, "/path/to/save/AnnData/object/")
```

The file "anndata_object.h5ad" will be saved into the user defined path and can then be used to perform the predictions. Beside the scRNA-seq data file, we also need to define a regulon database in form of a csv file with the coloumns "source", "target" and "weight". Within this package we provide the dorothea databases for human and mouse, downloaded from the decoupleR package. These files also contain the column "confidence" (levels A to D) with information on how well described a transcription factor and target gene interaction is in different resources. We recommend using the confidence levels A and B.

```{r, include = TRUE, eval = FALSE}
library(dplyr)

regulon <- read.csv(system.file("regulons", "human_dorothea_reg.csv", package = "LR2TF"))

```

## Usage
First the package and Seurat need to be imported:
&nbsp;

```{r eval = FALSE}
library(LR2TF)
library(Seurat)
```
&nbsp;

In the package an example dataset it included that can be accessed:
&nbsp;

```{r, include = TRUE, eval = FALSE}
data(bone_marrow_stromal_cell_example, package = "LR2TF")
```
&nbsp;

Before the execution, some parameters need to be defined:
&nbsp;

1. Path to save the results to
2. defining the confidence level of the dorothea regulon (A (strong) to E (low))
3. the organism of the sample was extracted from
4. Name of the meta-data field in the Seurat object of the cell annotation
5. Name of the meta-data field in the Seurat object of the condition annotation
6. List of condition comparison to be calculated

&nbsp;

```{r, include = TRUE, eval = FALSE}
out_path = '/Path/to/result/directory'
confidence_level = c("A", "B", "C", "D", "E")
organism = "human"
celltype_annotation = "annotation"
condition_annotation = "protocol"
comparison_list = list(c("control", "PMF,MF2"))
```
&nbsp;

After defining the necessary parameter the transcription factor activity can be predicted by calling:
&nbsp;

```{r, include = TRUE, eval = FALSE}
results = LR2TF::dorothea_tf_prediction(bone_marrow_stromal_cell_example, out_path, confidence_level, organism, condition_annotation, celltype_annotation, comparison_list)
```
&nbsp;

As result an object containing three fields is obtained (here example for control condition):
&nbsp;

* results[["control"]][["cluster"]] -> cluster specific transcription factors
* results[["control"]][["condition"]] -> condition specific transcription factors
* results[["control_average_expression"]] -> average gene expression matrix of specified condition

&nbsp;

For the conversion into the CrossTalkeR input format and the connection to receptor and ligands, either the cluster or the condition specific transcription factors can be used.
&nbsp;

```{r, include = TRUE, eval = FALSE}
result_table_control <- LR2TF::generate_CrossTalkeR_input_significant_table(results[["control"]][["condition"]], confidence_level, results[["control_average_expression"]])
result_table_pmf <- LR2TF::generate_CrossTalkeR_input_significant_table(results[["PMF_MF2"]][["condition"]], confidence_level, results[["PMF_MF2_average_expression"]])
```
&nbsp;

The last step is to combine previous results from ligand receptor interaction analyses (e.g. CellPhoneDB) with the transcription factor results:
&nbsp;

```{r, include = TRUE, eval = FALSE}
control_lr_file = "/Path/to/CellphoneDB/output/ControlLR.csv"
pmf_lr_file = "/Path/to/CellphoneDB/output/PMFLR.csv"

LR2TF::combine_LR_and_TF(result_table_control, control_lr_file, out_path, "control")
LR2TF::combine_LR_and_TF(result_table_pmf, pmf_lr_file, out_path, "PMF_MF2")

```
&nbsp;

## Analysis of mouse data
It is also possible to analyse scRNA-seq data from mouse samples. For this, the organism parameter must be set to "mouse" and the following function must be used when generating the CrossTalkeR input:
&nbsp;

```{r, include = TRUE, eval = FALSE}
LR2TF::generate_CrossTalkeR_input_mouse_significant_table()
```
&nbsp;

## References

[1] Garcia-Alonso L, Holland CH, Ibrahim MM, Turei D, Saez-Rodriguez J. “Benchmark and integration of resources for the estimation of human transcription factor activities.” Genome Research. 2019. DOI: 10.1101/gr.240663.118.

[2] A Valdeolivas, D Turei, A Gabor (2019) “OmnipathR: client for the OmniPath web service.” Bioconductor Package

[3] D Turei, T Korcsmaros and J Saez-Rodriguez (2016) OmniPath: guidelines and gateway for literature-curated signaling pathway resources. Nature Methods 13 (12); PMID: 27898060

[4] D Turei, A Valdeolivas, L Gul, N Palacio-Escat, M Klein, O Ivanova, M Olbei, A Gabor, F Theis, D Modos, T Korcsmaros and J Saez-Rodriguez (2021) Integrated intra- and intercellular signaling knowledge for multicellular omics analysis. Molecular Systems Biology 17: e9923; DOI: 10.15252/msb.20209923

[5] James S Nagai, Nils B Leimkühler, Michael T Schaub, Rebekka K Schneider, Ivan G Costa, CrossTalkeR: analysis and visualization of ligand–receptorne tworks, Bioinformatics, Volume 37, Issue 22, 15 November 2021, Pages 4263–4265, https://doi.org/10.1093/bioinformatics/btab370





#library(LR2TF)
library(Matrix)
library(Seurat)
#library(CrossTalkeR)
library(tibble)
library(dplyr)
library(org.Mm.eg.db)
library(org.Hs.eg.db)
library(clusterProfiler)
library(liana)
library(decoupleR)
library(reticulate)
devtools::load_all("/home/vanessa/repositories/CrossTalkeR/")
devtools::load_all("/home/vanessa/repositories/LR2TF/")
reticulate::use_condaenv("py39")
data(bone_marrow_stromal_cell_example, package = "LR2TF")
table_ctr <- read.csv("/home/vanessa/Projekte/CrossTalkeR/CTR_Testruns/decoupler_comparison/viper_olddorothea/LR/control_lr_ready.csv", row.names = 1)
table_exp <- read.csv("/home/vanessa/Projekte/CrossTalkeR/CTR_Testruns/decoupler_comparison/viper_olddorothea/LR/PMF,MF2_lr_ready.csv", row.names = 1)

parameters <- list("out_path" = "/home/vanessa/Projekte/CrossTalkeR/CTR_Testruns/decoupler_comparison/ulm_decouplerdorothea/",
                   reg = "/decoupler_comparison/wmean_decouplerdorothea/decoupler_doro_human.csv",
                   "organism" = "human",
                   "celltype" = "new_annotation",
                   "condition" = "protocol",
                   "comparison_list" = list(c("PMF,MF2", "control")),
                   "logfc" = 1,
                   "pval" = 0.05)


out_path <-  "/home/vanessa/Projekte/CrossTalkeR/CTR_Testruns/decoupler_comparison/ulm_decouplerdorothea/"
object_path <- "/home/vanessa/Projekte/CrossTalkeR/CTR_Testruns/decoupler_comparison/viper_olddorothea/bone_marrow_stromal_example_data.h5ad"
reg_path <- "/decoupler_comparison/wmean_decouplerdorothea/decoupler_doro_human.csv"

LR2TF::run_py_decoupler_ulm(object_path, reg_path, out_path)

results <- LR2TF::tf_activity_analysis(seuratobject = bone_marrow_stromal_cell_example,
                                       tf_activities = "/home/vanessa/Projekte/CrossTalkeR/CTR_Testruns/decoupler_comparison/ulm_decouplerdorothea/decoupler_results.csv",
                                       arguments_list = parameters)

ctr_inptu <- LR2TF::combine_LR_and_TF(results@CTR_input_condition[["control"]], table_ctr, parameters$out_path, "control")
exp_input <- LR2TF::combine_LR_and_TF(results@CTR_input_condition[["PMF_MF2"]], table_exp, parameters$out_path, "PMF_MF2")



